### DISCUSSION TITLE
Input capture?

### DISCUSSION THREAD

**Discussion by zoenetic** [niri-wm/niri#821] [Ideas]
Hi all; I'm using niri on nixos, and trying to use deskflow to share my keyboard/mouse with another device. When I try to run the deskflow server it fails with the error 'failed to initialize input capture session' and my googling indicates this might be because an input capture portal isn't available.

My niri-portals.conf file contains the following:

```
[preferred]
default=gnome;gtk;
org.freedesktop.impl.portal.Access=gtk;
org.freedesktop.impl.portal.Secret=gnome-keyring;
```

I'm guessing this means it'll fall back to using the gnome portal for protocols not listed (and I think the gnome portal supports input capture) but I'm not sure.

Would you expect input capture to work with niri? If so, any advice on getting it working?

**Comment by YaLTeR** [comment]
Hello, it is in fact not implemented yet. The gnome portal may support input capture, but it still needs to be implemented in niri.

**Reply by zoenetic** [reply]
Aha! Okay no problem, and thanks for the prompt response. I guess you can consider this a feature request then haha

**Reply by YaLTeR** [reply]
Made an issue: https://github.com/YaLTeR/niri/issues/823

**Reply by BlueGuitarCL** [reply]
Quick bump-up! Loving niri so far, daily driving it since Dec 2024. Tho just tried using a Deskflow [Fork](https://github.com/input-leap/input-leap) got me into the same issue.
```
[Wed Jan 22] blu@a5rch:~ $ input-leaps --name a5rch -d DEBUG -c ~/.config/InputLeap/input-leap.conf --no-daemon[2025-01-22T17:59:26] DEBUG: opening configuration "/home/blu/.config/InputLeap/input-leap.conf"
[2025-01-22T17:59:26] DEBUG: configuration read successfully
[2025-01-22T17:59:26] DEBUG: adopting new buffer
[2025-01-22T17:59:26] DEBUG: adopting new buffer
[2025-01-22T17:59:26] DEBUG: opened display
[2025-01-22T17:59:26] DEBUG: GLib thread running
[2025-01-22T17:59:26] DEBUG: Setting up the InputCapture session
[2025-01-22T17:59:26] DEBUG: Session ready
[2025-01-22T17:59:26] ERROR: Failed to initialize InputCapture session, quitting: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface “org.freedesktop.portal.InputCapture” on object at path /org/freedesktop/portal/desktop
started server (IPv4), waiting for clients
[2025-01-22T17:59:26] DEBUG: event queue is ready
[2025-01-22T17:59:26] DEBUG: add pending events to buffer
[2025-01-22T17:59:26] DEBUG: Shutting down GLib thread
[2025-01-22T17:59:26] DEBUG: add pending events to buffer
[2025-01-22T17:59:26] DEBUG: event queue is ready
[2025-01-22T17:59:26] DEBUG: screen "a5rch" shape changed
[2025-01-22T17:59:26] DEBUG: closed display
[2025-01-22T17:59:26] DEBUG: adopting new buffer
[2025-01-22T17:59:26] DEBUG: discarding 1 event(s)
[2025-01-22T17:59:26] NOTE: stopped server
```
Anyways, keep it up :). Rn im happy with the new floating windows feature, so im patient enough for this to go out on a release.

P.D.: Gonna take the opportunity to add an idea i been having around niri. I need `wlr-virtual-pointer-unstable-v1` implemented in order to use things like `wayvnc` to control my linux machine thru' other PCs. Tho, for my typical usecase (accesing though mobile devices on touchscreens), tiling WM's/wlCompositors falls short. So, im wondering if it should be a good idea adding Focus Left/Right/WorkspaceUp|Down on window decorations as buttons in order to make the touchscreen experience usable. (I know a workaround could be using `niri msg`'s for doing the focus stuff, but i don't like the idea of having several waybar instances with those buttons, because i want to keep the window aspect ratios to match my screens)

Just spilled the idea here bc' i don't feel like the idea is polished enough yet. But feel free to answer or make a separate discussion/issue for the topic :D

**Reply by YaLTeR** [reply]
> Quick bump-up!

Make sure to +1 the issue so it shows up when sorting by reactions

> I need wlr-virtual-pointer-unstable-v1

Recently merged to main actually

> adding Focus Left/Right/WorkspaceUp|Down on window decorations

Well we don't really have much in terms of window decorations now so there's nowhere to add these.

> in order to make the touchscreen experience usable

Also #372 may help

**Comment by selckin** [comment]
https://github.com/feschber/lan-mouse works

**Reply by gitjuice** [reply]
> https://github.com/feschber/lan-mouse works

can you please explain how? I do not receive any connection when trying to add a client

**Reply by AkiraLink** [reply]
> > https://github.com/feschber/lan-mouse works
> 
> can you please explain how? I do not receive any connection when trying to add a client

have you found a solution?? I am in dire need of anything that work like DeskFlow, be it lan mouse or something else

**Reply by gitjuice** [reply]
> > > https://github.com/feschber/lan-mouse works
> > 
> > 
> > can you please explain how? I do not receive any connection when trying to add a client
> 
> have you found a solution?? I am in dire need of anything that work like DeskFlow, be it lan mouse or something else

unfortunately no, I hope they will figure out a solution soon. There's already a draft PR (check the comment below), so I'm confident deskflow will be supported in the near future.

**Reply by drewofdoom** [reply]
FYI for everyone finding this, I tried lan-mouse and did get it to work. Here's my setup:

I'm on Bluefin (technically my own ublue build downstream from Bluefin) with greetd, niri, and noctalia.

First tried the version from Fedora Terra repo, which is currently on 0.10.0 -  That seemed like it would have worked if the clients weren't having issues talking to each other.

Pulling down the binary build from the lan-mouse release, I was able to launch it, set up the connection, and get it going. It detects the input portal is not available and falls back to wlroots and layer-shell. After that, it really just kinda works.

**Comment by mrnicegyu11** [comment]
Linking this PR here which might allow `input-leap` to work if it gets merged, or as far as I understand brings the basic requirements for this at least: https://github.com/YaLTeR/niri/pull/1966

**Comment by jeanlescure** [comment]
## GOOD NEWS! We have a workaround :rocket:

**TL;DR**

Launch Deskflow in X11 mode: 

```
env XDG_SESSION_TYPE=x11 DISPLAY=:1 deskflow
```

or like this if running Flatpak version:

```
flatpak run --env=XDG_SESSION_TYPE=x11 --env=DISPLAY=:1 org.deskflow.deskflow
```

You know this is working when you see `INFO: using legacy x windows screen` in Deskflow's logs

Configure hotkeys in Deskflow to switch screens (instead of mouse-to-edge)

Make the Deskflow window fullscreen in Niri before switching

That's it. Keyboard, mouse, and clipboard sharing all "work" :partying_face:

(at least while Niri team gets this sorted out :smile:)

---

Full context for the Niri team:

**Setup:**

- Niri v25.11 on Zorin OS 18 (Ubuntu 24.04 base)
- Deskflow (Flatpak) as server, MacBook as client
- `xwayland-satellite` running on `:1`
- **Big** fan of Niri, it's making my life so much easier by knowing exactly where all my vscode windows are at :heart:

**The problem (same as everyone here):**

Deskflow in Wayland/ei mode fails with `CreateSession` failed because `xdg-desktop-portal-gnome`'s `InputCapture` implementation depends on Mutter internals that don't exist in Niri

**Reason the workaround works:**

Setting` XDG_SESSION_TYPE=x11` forces Deskflow to use its `legacy x windows` screen path instead of `ei screen` for wayland

(that's literally the only env var Deskflow checks, see [source](https://github.com/deskflow/deskflow/blob/master/src/lib/common/PlatformInfo.h))

This makes it run through `xwayland-satellite`, bypassing the broken `InputCapture` portal entirely

Hotkeys are needed because screen-edge detection requires `InputCapture`, which is the whole problem

Fullscreen is critical because without it, niri consumes horizontal pointer movement for its column scrolling behavior. The client only receives vertical mouse input

Going fullscreen disables niri's scrolling interception and `xwayland-satellite` gets both axis

**What works with this workaround:**

- Keyboard and mouse sharing via hotkey switching
- Mouse movement from client back to server (edge detection works in that direction)
- Clipboard sharing

**What still is a pain:**

- Deskflow window **MUST** be fullscreen **AND** focused for the hotkey to work
- Mouse-to-edge switching from server to client only **randomly** works (most of the time no `InputCapture`, as expected)
- Horizontal mouse input when Deskflow isn't fullscreen (niri's scrolling intercepts it)

**Note for Niri devs:**

The horizontal axis issue suggests input routing to `xwayland-satellite` clients isn't axis-neutral in tiled/floating mode. Niri appears to intercept horizontal pointer movement for column scrolling even when the focused window should be receiving all input. Fullscreen bypasses this. Might be worth looking at independently of the `InputCapture` work.

----

Edit: added non-flatpak command to launch deskflow, and a few more details

**Reply by airtonix** [reply]
Cheers, this works for Synergy3 too. 

I just keep a Synergy3 window fullscreened and alt tab to it then the screen edges works. I have copy paste working (but i can't be sure if that's kdeconnect or synergy)


```sh
/usr/bin/flatpak run \
  --env=XDG_SESSION_TYPE=x11 \
  --env=DISPLAY=:1 \
  --branch=master \
  --arch=x86_64 \
  --command=electron-wrapper \
  --file-forwarding \
  com.symless.synergy \
  @@u %U @@
```

What doesn't work, is that when my mouse is across the network on the windows machine: 

- alt tab makes niri do alt tab things
- super + left click also makes niri do window grab things
- basically... anything you have a keyboard shortcut for in niri... yep. niri wins.

I dont have a niri short cut for super + right drag, so altSnap in windows gets this and i can resize window frames in windows with that.

**Reply by jeanlescure** [reply]
> anything you have a keyboard shortcut for in niri... yep. niri wins.

+1 on this behavior, a bit annoying but I'm already getting muscle memory of doing alt+tab on the client keyboard when working on the client screen 😄

**Comment by danremollino** [comment]
Thanks @jeanlescure. Following your instructions I was able to have my macOS client connect to my CachyOS server. However, I can't get the mouse/keyboard to switch computers via hotkey.

When you say to make sure Deskflow is full screen in Niri, do you mean using a Niri hotkey to `fullscreen-window` on it? I'm currently doing this with no luck.

**Reply by jeanlescure** [reply]
Hi @danremollino, I use `Mod+f` to make the Deskflow window go fullscreen, then with Deskflow focused on my Zorin OS box I can use my hotkey to switch to my macOS client

I can say it's _not_ ideal for it to be a hard requirement for Deskflow to be the focused window, specially when I'm doing research on my Zorin laptop and want to read that research while I'm working on my Macbook

Sometimes I'll just save my research to a file and open it via smb shared dir on my Macbook

Other times I just switch session from my Niri session to the default Zorin OS Wayland session

For now this works for me, but I do hope some day an `InputCapture` fix is implemented in Niri

And a side-note on that, at some point I did consider submitting a PR, but seeing as there are 151 PRs (currently), I'd rather give 
@YaLTeR some breathing room to deal with them before adding yet another one

Also, @YaLTeR, if you'd like some help going through PRs and labeling them to prioritize, my DMs are open! 😄

**Reply by danremollino** [reply]
Hmm. Did some more testing it seems I can't get Deskflow to execute ANY kb shortcuts that I define in the Configure Server dialog box. Tried with while in focus and windowed, maximized to the viewport, and fullscreened.

I'm using Deskflow from the Arch extra repo. Maybe I'll give the Flatpak a shot as I think you are using that.

**EDIT:** No luck on the flatpak. But I think [this](https://github.com/deskflow/deskflow/issues/7596) is the kb issue.

**Comment by MGSousa** [comment]
Since niri yet doesn't implement the RemoteDesktop/InputCapture XDG portal interfaces (tracked in #390 and #823), I ended up using waynergy (https://github.com/r-c-f/waynergy) as a drop-in replacement client instead of the XDG_SESSION_TYPE=x11 + xwayland-satellite workaround mentioned above.

This is a valid, faster alternative since it sidesteps the portal gap entirely rather than working around it until the PRs above are merged.

Waynergy speaks the Synergy/Barrier wire protocol directly (so it connects to an existing `Deskflow/Barrier/InputLeap/Synergy` server completely unmodified) and injects input via the standard `zwlr_virtual_pointer_manager_v1` / `zwp_virtual_keyboard_manager_v1` Wayland protocols, which niri already has.

A couple of things worth mentioning:

- If your server is **Windows** or **macOS**, you'll likely need a custom xkb_keymap (waynergy ships doc/xkb/keycodes/win/ doc/xkb/keycodes/mac for exactly this); the default raw/keysym modes can otherwise scramble specific keys. The Windows Server recipe that worked for me matches the issue [#15](https://github.com/r-c-f/waynergy/issues/15) in the waynergy repo.

- Same underlying limitation as the InputCapture gap discussed here applies to Mod/Super-based compositor shortcuts: virtual-keyboard injected key events go straight to the focused client's wl_keyboard (confirmed by checking `virtual_keyboard_handle.rs`), bypassing niri's own keybinding matcher entirely. 
  - Regular typing/mouse works fine, but `Mod+<X>` style mapping of niri shortcuts doesn't fire through this path, on niri, which is inherent to how the protocol is implemented, not something waynergy or niri config can fix. (I forked it with a small patch (more oriented for niri, in this case) which watches for your chosen modifier + keysym in the client and spawns `niri msg action <action>` which can be set on the `config.ini` file) - https://github.com/MGSousa/waynergy
 
config.in:
```ini
host = X.X.X.X
port = YYYY
...

[niri-hotkey]
q = niri msg action close-window
t = niri msg action spawn -- <application_bin_name>
....
```
